yt/frontends/nc4_cm1/data_structures.py

```diff
@@ -0,0 +1,233 @@
+
+    def _populate_grid_objects(self):
+        # For each grid g, this must call:
+        #   g._prepare_grid()
+        #   g._setup_dx()
+        # This must also set:
+        #   g.Children <= list of child grids
+        #   g.Parent   <= parent grid
+        # This is handled by the frontend because often the children must be
+        # identified.
+        self.grids = np.empty(self.num_grids, dtype="object")
+        for i in range(self.num_grids):
+            g = self.grid(i, self, self.grid_levels.flat[i], self.grid_dimensions[i])
+            g._prepare_grid()
+            g._setup_dx()
+            self.grids[i] = g
+
+
+class CM1Dataset(Dataset):
+    _index_class = CM1Hierarchy
+    _field_info_class = CM1FieldInfo
+
+    def __init__(
+        self, filename, dataset_type="cm1", storage_filename=None, units_override=None
+    ):
+        self.fluid_types += ("cm1",)
+        self._handle = NetCDF4FileHandler(filename)
+        # refinement factor between a grid and its subgrid
+        self.refine_by = 1
+        super(CM1Dataset, self).__init__(
+            filename, dataset_type, units_override=units_override, unit_system="mks"
+        )
+        self.storage_filename = storage_filename
+        self.filename = filename
+
+    def _setup_coordinate_handler(self):
+        super(CM1Dataset, self)._setup_coordinate_handler()
+        self.coordinates._x_pairs = (("x", "y"), ("y", "x"), ("z", "x"))
+        self.coordinates._y_pairs = (("x", "z"), ("y", "z"), ("z", "y"))
+
+    def _set_code_unit_attributes(self):
+        # This is where quantities are created that represent the various
+        # on-disk units.  These are the currently available quantities which
+        # should be set, along with examples of how to set them to standard
+        # values.
+        with self._handle.open_ds() as ds:
+            length_unit = ds.variables["xh"].units
+        self.length_unit = self.quan(1.0, length_unit)
+        self.mass_unit = self.quan(1.0, "kg")
+        self.time_unit = self.quan(1.0, "s")
+        self.velocity_unit = self.quan(1.0, "m/s")
+        self.time_unit = self.quan(1.0, "s")
+
+    def _parse_parameter_file(self):
+        # This needs to set up the following items.  Note that these are all
+        # assumed to be in code units; domain_left_edge and domain_right_edge
+        # will be converted to YTArray automatically at a later time.
+        # This includes the cosmological parameters.
+        #
+        #   self.unique_identifier      <= unique identifier for the dataset
+        #                                  being read (e.g., UUID or ST_CTIME)
+        self.unique_identifier = int(os.stat(self.parameter_filename)[stat.ST_CTIME])
+        self.parameters = {}  # code-specific items
+        with self._handle.open_ds() as ds:
+            dims = [ds.dimensions[i].size for i in ["xh", "yh", "zh"]]
+            coords = {i: ds.variables[i][:] for i in ["xh", "yh", "zh"]}
```

Comment:
looks like this isn't used before line 169. I suggest merging those

Reply:
I ended up removing `coords` entirely -- I'm not sure why we were copying the coordinate data over to parameters as it's not used elsewhere and we can extract it from the yt dataset after loading.